server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java

```diff
@@ -0,0 +1,196 @@
+                        response.append("myself,");
+                     }
+                  } else {
+                     // We could not retrieve information from the node.
+                     // We add the information as if was disconnected.
+                     response.append(member).append(' ');
+                     response.append(member).append(':').append(0).append('@').append(0);
+                     response.append(",,shard-id=").append(member).append(' ');
+
+                     if (isMyself) {
+                        response.append("myself,");
+                     } else {
+                        response.append("fail?,");
+                     }
+                     health = "disconnected";
+                  }
+
+                  response.append("master").append(' ');
+                  response.append('-').append(' ');
+                  response.append('0').append(' ');
+                  response.append(Instant.now().getEpochSecond()).append(' ');
+                  response.append(topology.getTopologyId()).append(' ');
+                  response.append(health).append(' ');
+                  serializeSegments(response, owner);
+               }
+               return "$" + response.length() + CRLF + response + CRLF;
+            });
+   }
+
+   private static void serializeSegments(StringBuilder response, IntSet ranges) {
+      BitSet bitSet = BitSet.valueOf(ranges.toBitSet());
+      boolean first = true;
+      for (int i = bitSet.nextSetBit(0); i >= 0; i = bitSet.nextSetBit(i + 1)) {
+         if (!first) {
+            response.append(' ');
+         }
+         first = false;
+         int runStart = i;
+         while (bitSet.get(i + 1)) {
+            i++;
+         }
+         response.append(runStart).append('-').append(i);
+      }
+   }
+
+   private static CompletionStage<Map<Address, String>> readNodeInformation(List<Address> members, ClusterExecutor executor) {
+      final Map<Address, String> response = new ConcurrentHashMap<>(members.size());
+      return executor.filterTargets(members)
+            .submitConsumer(NODES::readLocalNodeInformation, (address, res, t) -> {
+               if (t != null) {
+                  throw CompletableFutures.asCompletionException(t);
+               }
+               response.put(address, res);
+            }).thenApply(ignore -> response);
+   }
+
+   private static String readLocalNodeInformation(EmbeddedCacheManager ecm) {
+      CacheManagerInfo info = ecm.getCacheManagerInfo();
+      String name = info.getNodeName();
+      Address address = findPhysicalAddress(ecm);
+      int port = findPort(address);
+      String addressString = address != null ? getOnlyIp(address) : info.getNodeAddress();
+
+      StringBuilder sb = new StringBuilder();
+      sb.append(name).append(' ');
+      sb.append(addressString).append(':').append(port).append('@').append(port);
+      sb.append(",,shard-id=").append(name).append(' ');
```

Comment:
Is the space at the end always needed?

Reply:
Yeah, in this format, they use space as a separator.

Reply:
Okay, I was thinking it was a bit of a waste if it was always there, even for the last node. But if that is how they want it :)